src/main/java/newbank/client/ExampleClient.java

```diff
@@ -6,77 +6,102 @@
 import java.io.PrintWriter;
 import java.io.Reader;
 import java.net.Socket;
+import java.net.SocketException;
 import java.net.UnknownHostException;
 
 import newbank.utils.ConsoleDisplay;
 import newbank.utils.Display;
 
 public class ExampleClient extends Thread {
 
-  private Socket server;
-  private PrintWriter bankServerOut;
-  private BufferedReader userInput;
-  private Thread bankServerResponceThread;
+  public static final int DEFAULT_PORT = 14002;
+  public static final String DEFAULT_IP = "localhost";
+  private final Socket server;
+  private final PrintWriter bankServerOut;
+  private final BufferedReader userInput;
+  private final Thread bankServerResponseThread;
   private Display display = new ConsoleDisplay();
 
+  /**
+   * @param ip an ip address, or the loopback address
+   * @param port the port that the newbank server will be listening on
+   * @throws ConfigurationException af the server cannot be reached at the supplied ip and port
+   * @throws IOException if an I/O exception occurs while trying to connect to the server
+   */
   public ExampleClient(String ip, int port, Reader reader)
-      throws UnknownHostException, IOException {
-    server = new Socket(ip, port);
+      throws IOException, IllegalThreadStateException, ConfigurationException {
+    server = getSocket(ip, port);
     userInput = new BufferedReader(reader);
     bankServerOut = new PrintWriter(server.getOutputStream(), true);
 
-    bankServerResponceThread =
+    bankServerResponseThread =
         new Thread() {
-          private BufferedReader bankServerIn =
+          private final BufferedReader bankServerIn =
               new BufferedReader(new InputStreamReader(server.getInputStream()));
 
           public void run() {
-            try {
-              while (true) {
-                String response = bankServerIn.readLine();
-
-                if (response == null) {
-                  // socket broken, end thread
-                  return;
-                }
-
-                display.writeLine(response);
+            while (true) {
+              String response = null;
+              try {
+                response = bankServerIn.readLine();
+              } catch (IOException e) {
+                e.printStackTrace();
+                Thread.currentThread().interrupt();
```

Comment:
This line looks suspicious. A "return" is all we needed?

Jules

Reply:
Ah, because the thread has no parent, so nobody needs to be notified of the interrupt!
I gotchya. Still getting used to Java's threads.

Reply:
If we change the line with the "interrupt" back to a return as before then the run() method hasn't changed? and perhaps needs to be back to its original form.

Kindest Regards,
Jules